src/composeappmanager.cc

```diff
@@ -145,25 +227,274 @@ data::InstallationResult ComposeAppManager::install(const Uptane::Target &target
+  return download_ctx->write(data, (buf_size * buf_numb));
+}
+
+void RegistryClient::downloadBlob(const std::string& repo, const HashedDigest& digest,
+                                  const boost::filesystem::path& file_path, size_t expected_size) const {
+  auto compose_app_blob_url{conf_.repo_base_url + repo + "/blobs/" + digest()};
+
+  LOG_DEBUG << "Downloading App blob: " << compose_app_blob_url;
+
+  std::vector<std::string> registry_repo_request_headers{getBearerAuthHeader(repo)};
+  auto registry_repo_client{http_client_factory_(&registry_repo_request_headers)};
+
+  std::ofstream output_file{file_path.string(), std::ios_base::out | std::ios_base::binary};
+  if (!output_file.is_open()) {
+    throw std::runtime_error("Failed to open a file: " + file_path.string());
+  }
+  MultiPartSHA256Hasher hasher;
+
+  DownloadCtx download_ctx{output_file, hasher, expected_size};
+  auto get_blob_resp = registry_repo_client->download(compose_app_blob_url, DownloadHandler, nullptr, &download_ctx, 0);
+
+  if (!get_blob_resp.isOk()) {
+    throw std::runtime_error("Failed to download App blob: " + get_blob_resp.getStatusStr());
+  }
+
+  output_file.close();
+  std::size_t recv_blob_file_size{download_ctx.written_size};
+
+  if (recv_blob_file_size != expected_size) {
+    std::remove(file_path.c_str());
+    throw std::runtime_error(
+        "Size of downloaded App blob does not equal to "
+        "the expected one: " +
+        std::to_string(recv_blob_file_size) + " != " + std::to_string(expected_size));
+  }
+
+  auto recv_blob_hash{boost::algorithm::to_lower_copy(hasher.getHexDigest())};
+
+  if (recv_blob_hash != digest.hash()) {
+    std::remove(file_path.c_str());
+    throw std::runtime_error(
+        "Hash of downloaded App blob does not equal to "
+        "the expected one: " +
+        recv_blob_hash + " != " + digest.hash());
+  }
+}
+
+std::string RegistryClient::getBasicAuthHeader() const {
+  LOG_DEBUG << "Getting Docker Registry credentials from " << conf_.auth_creds_endpoint;
+
+  auto creds_resp = ota_lite_client_->get(conf_.auth_creds_endpoint, AuthMaterialMaxSize);
+
+  if (!creds_resp.isOk()) {
+    throw std::runtime_error("Failed to get Docker Registry credentials from " + conf_.auth_creds_endpoint +
+                             "; error: " + creds_resp.getStatusStr());
+  }
+
+  auto creds_json = creds_resp.getJson();
+  auto username = creds_json["Username"].asString();
+  auto secret = creds_json["Secret"].asString();
+
+  if (username.empty() || secret.empty()) {
+    throw std::runtime_error("Got invalid Docker Registry credentials: " + creds_resp.body);
+  }
+
+  std::string auth_secret_str{username + ':' + secret};
+  auto encoded_auth_secret = Utils::toBase64(auth_secret_str);
+
+  LOG_DEBUG << "Got Docker Registry credentials, username: " << username;
+  return "authorization: basic " + encoded_auth_secret;
+}
+
+std::string RegistryClient::getBearerAuthHeader(const std::string& repo) const {
```

Comment:
this might be too painful for now and we should just keep it like this with a TODO. However, docker provides a way to find credentials. In theory, on LmP, you look at `/usr/lib/docker/config.json` and then `$HOME/.docker/config.json` for `auths` and `credHelpers`. If "auths" you just use the value provided. If "credsHelpers", you call that program and use its stdout.

Maybe we assert hub.foundries.io for now though and because its hub.fio, we know how to handle this w/o using the credsHelper.

Reply:
The credHelpers relies on aktaulizr-get. What's the point to go from aktualizr to the docker stuff which in turn comes back to aktualizr if 'we' are aktualizr and go straight to the auth service.
It's not `hub.foundries.io`, it's `https://ota-lite.foundries.io:8443/hub-creds/`.

Reply:
what i was trying to say here is that: we should really make this generic and not tied to hub.foundries.io. maybe eventually though since its a bit more work. in the case of hub.foundries.io, we could skip looking up the cred helper since we know how to do that.

what's nice about the docker-compose implementation is that it could download the archive from *any* docker registry with *any* credentials docker supports. 

but since this is tied to hub.foundries.io, I think we can skip the configuration data and just hard-code to hub.foundries.io and raise an error if its not a hub.foundries.io app. In the future we could add logic to support other docker registries.

Reply:
I see what you mean. Still, we can change our registry's hostname overtime or deploy additional one with the same way of getting creds to it (via https://ota-lite.foundries.io:8443/hub-creds).